### PR TITLE
SONAR-7196 WS api/measures/component_tree supports sorting by period variation

### PR DESCRIPTION
**Pull request by teryk**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentTreeSort.java [hunk 1]
```diff
@@ -158,6 +193,29 @@ public Double apply(@Nonnull ComponentDtoWithSnapshotId input) {
...
+
+    private ComponentDtoWithSnapshotIdToMeasureVariationValue(@Nullable MetricDto metric,
+      Table<String, MetricDto, MeasureDto> measuresByComponentUuidAndMetric, int variationIndex) {
+      this.metric = metric;
+      this.measuresByComponentUuidAndMetric = measuresByComponentUuidAndMetric;
+      this.variationIndex = variationIndex;
+    }
+
+    @Override
+    public Double apply(@Nonnull ComponentDtoWithSnapshotId input) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentTreeSort.java]: MAJOR The "input" parameter nullability is different in the superclass method, and that should not be changed. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2638)

File: server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentTreeAction.java [hunk 2]
```diff
@@ -250,11 +258,14 @@ private static ComponentTreeWsRequest toComponentTreeWsRequest(Request request)
       "The '%s' parameter must have at least %d characters", Param.TEXT_QUERY, QUERY_MINIMUM_LENGTH);
     String metricSortValue = componentTreeWsRequest.getMetricSort();
     checkRequest(!componentTreeWsRequest.getMetricKeys().isEmpty(), "The '%s' parameter must contain at least one metric key", PARAM_METRIC_KEYS);
-    checkRequest(metricSortValue == null ^ componentTreeWsRequest.getSort().contains(METRIC_SORT),
-      "To sort by a metric, the '%s' parameter must contain '%s' and a metric key must be provided in the '%s' parameter",
-      Param.SORT, METRIC_SORT, PARAM_METRIC_SORT);
+    checkRequest(metricSortValue == null ^ componentTreeWsRequest.getSort().contains(METRIC_SORT)
+      ^ componentTreeWsRequest.getSort().contains(METRIC_PERIOD_SORT),
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentTreeAction.java]: BLOCKER NullPointerException might be thrown as 'getSort' is nullable here [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2259)

File: server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentTreeAction.java [hunk 3]
```diff
@@ -250,11 +258,14 @@ private static ComponentTreeWsRequest toComponentTreeWsRequest(Request request)
       "The '%s' parameter must have at least %d characters", Param.TEXT_QUERY, QUERY_MINIMUM_LENGTH);
     String metricSortValue = componentTreeWsRequest.getMetricSort();
     checkRequest(!componentTreeWsRequest.getMetricKeys().isEmpty(), "The '%s' parameter must contain at least one metric key", PARAM_METRIC_KEYS);
-    checkRequest(metricSortValue == null ^ componentTreeWsRequest.getSort().contains(METRIC_SORT),
-      "To sort by a metric, the '%s' parameter must contain '%s' and a metric key must be provided in the '%s' parameter",
-      Param.SORT, METRIC_SORT, PARAM_METRIC_SORT);
+    checkRequest(metricSortValue == null ^ componentTreeWsRequest.getSort().contains(METRIC_SORT)
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentTreeAction.java]: BLOCKER NullPointerException might be thrown as 'getSort' is nullable here [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2259)

File: server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentTreeAction.java [hunk 4]
```diff
@@ -250,11 +258,14 @@ private static ComponentTreeWsRequest toComponentTreeWsRequest(Request request)
...
+      ^ componentTreeWsRequest.getSort().contains(METRIC_PERIOD_SORT),
+      "To sort by a metric, the '%s' parameter must contain '%s' or '%s', and a metric key must be provided in the '%s' parameter",
+      Param.SORT, METRIC_SORT, METRIC_PERIOD_SORT, PARAM_METRIC_SORT);
     checkRequest(metricSortValue == null ^ componentTreeWsRequest.getMetricKeys().contains(metricSortValue),
       "To sort by the '%s' metric, it must be in the list of metric keys in the '%s' parameter", metricSortValue, PARAM_METRIC_KEYS);
+    checkRequest(componentTreeWsRequest.getMetricPeriodSort() == null ^ componentTreeWsRequest.getSort().contains(METRIC_PERIOD_SORT),
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentTreeAction.java]: BLOCKER NullPointerException might be thrown as 'getSort' is nullable here [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2259)

File: server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentTreeDataLoader.java [hunk 5]
```diff
@@ -254,6 +255,10 @@ private static void addBestValuesToMeasures(Table<String, MetricDto, MeasureDto>
       .toList();
   }
 
+  private static boolean isSortByMetric(ComponentTreeWsRequest wsRequest) {
+    return wsRequest.getSort().contains(METRIC_SORT) || wsRequest.getSort().contains(METRIC_PERIOD_SORT);
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentTreeDataLoader.java]: BLOCKER NullPointerException might be thrown as 'getSort' is nullable here [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2259)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* BLOCKER 4 blocker
* MAJOR 1 major

Watch the comments in this conversation to review them.

**Comment by julienlancelot** [comment]
Please clarify unit tests of ComponentTreeSortTest, and fix NPE.

**Pull request closed without merging**